### ISSUE TITLE
How to convert from colmap coordinate system to Unity coordinate system?

### ISSUE THREAD

**Issue by adammpolak** [colmap/colmap#1377]
I created a reconstruction but when I type in the locations of the images using quaternion (QW, QX, QY, QZ) and a translation vector (TX, TY, TZ) they are not in the same locations as they are in the reconstruction.

**Comment by tsattler** [comment]
Colmap is using the coordinate system commonly used in the computer vision literature (see tutorial and documentation). I'd assume that Unity is using the coordinate system commonly used in computer graphics. If this is the case, you will need to convert from one to the other.

**Comment by adammpolak** [comment]
> Colmap is using the coordinate system commonly used in the computer vision literature (see tutorial and documentation). I'd assume that Unity is using the coordinate system commonly used in computer graphics. If this is the case, you will need to convert from one to the other.

Really appreciate the response thank you!

Unfortunately I have ctl+f every page of the colmap documentation searching for "axis" "axes" and "orientation" and have not found any indication about their orientation :(

**Comment by tsattler** [comment]
The local camera coordinate system used by Colmap is described here: https://colmap.github.io/format.html#images-txt

> The local camera coordinate system of an image is defined in a way that the X axis points to the right, the Y axis to the bottom, and the Z axis to the front as seen from the image.

**Comment by adammpolak** [comment]
> 

Yes thank you! I was able to find the reference to the camera coordinates, but what I am looking for is the **orientation of the global coordinates** (6dof camera poses relative to each other).

**Comment by tsattler** [comment]
There is no unique definition of the global coordinates. Structure-from-Motion reconstructions such as those computed by Colmap are defined up to an arbitrary rotation, translation, and uniform scaling, i.e., rotating, translating, and scaling the scene will result in the same reprojections. Colmap initializes the global coordinate system of the model as the local coordinate system of the first image that gets reconstructed, but the global coordinate system changes as part of the bundle adjustment process.

**Comment by adammpolak** [comment]
Really appreciate the response!

Sorry for being confusing regarding my questions, I will try to be more clear

Explanation of how right-handed vs. left-handed coordinate systems:
> Typically, 3-D graphics applications use two types of Cartesian coordinate systems: left-handed and right-handed. In both coordinate systems, the positive x-axis points to the right, and the positive y-axis points up. You can remember which direction the positive z-axis points by pointing the fingers of either your left or right hand in the positive x direction and curling them into the positive y direction. The direction your thumb points, either toward or away from you, is the direction that the positive z-axis points for that coordinate system. The following illustration shows these two coordinate systems.
> ![image](https://user-images.githubusercontent.com/12940507/149672502-2b2213b1-ed69-43a0-882c-860cf41f97d0.png)

When choosing the x y and z axis that is an "orientation". These orientations are sometimes referred to as "right handed" or "left handed" based on the order of the axis. This affects how you would perform translations and rotations based on what axis you would want to rotate around.

It is hard for me to believe that colmap arbitrarily chooses orientation on every reconstruction (unless that is what you are telling me!).

Colmap provides x, y, z translations and rotations as output. This means the colmap is deciding arbitrarily (but deterministically) what axis is x y and z. These decisions can be seen in the GUI by the different color axis.

My question is what is the orientation of the deterministic decision making? Is it left handed or right handed, and what would the axis be in these reconstructions? Knowing this I could then transform/rotate to fit other systems (such as Unity).

**Comment by tsattler** [comment]
Colmap uses a right-handed coordinate system as the local camera coordinate system of each camera: positive x is pointing to the right, positive y is pointing down, positive z is pointing away from the camera towards the part of the scene that is visible in the camera (see https://colmap.github.io/format.html#images-txt).

As mentioned before, Colmap initializes the world coordinate system by the local camera coordinate system of the first image that gets reconstructed. The choice of the first image is not necessarily deterministic as some parts of Colmap (relative and absolute pose estimation) are based on randomized algorithms (RANSAC to be more precise).
